src/transport.js

```diff
@@ -34,59 +41,86 @@ module.exports = function (swarm) {
       // b) if multiaddrs.length = 1, return the conn from the
       // transport, otherwise, create a passthrough
       if (multiaddrs.length === 1) {
-        const conn = t.dial(multiaddrs.shift(), {ready: () => {
-          const cb = callback
-          callback = noop // this is done to avoid connection drops as connect errors
-          cb(null, conn)
-        }})
-        conn.once('error', () => {
-          callback(new Error('failed to connect to every multiaddr'))
+        const conn = t.dial(multiaddrs.shift())
+
+        conn.once('error', connectError)
+
+        conn.on('connect', () => {
+          conn.removeListener('error', connectError)
+          callback(null, conn)
         })
+
         return conn
       }
+      function connectError () {
+        callback(new Error('failed to connect to every multiaddr'))
+      }
 
       // c) multiaddrs should already be a filtered list
       // specific for the transport we are using
-      const pt = new Duplexify()
+      const proxyConn = new Connection()
 
       next(multiaddrs.shift())
-      return pt
+
+      return proxyConn
+
+      // TODO improve in the future to make all the dials in paralell
       function next (multiaddr) {
-        const conn = t.dial(multiaddr, {ready: () => {
-          pt.setReadable(conn)
-          pt.setWritable(conn)
-          pt.getObservedAddrs = conn.getObservedAddrs.bind(conn)
-          const cb = callback
-          callback = noop // this is done to avoid connection drops as connect errors
-          cb(null, pt)
-        }})
-
-        conn.once('error', () => {
+        const conn = t.dial(multiaddr)
+
+        conn.once('error', connectError)
+
+        function connectError () {
           if (multiaddrs.length === 0) {
             return callback(new Error('failed to connect to every multiaddr'))
           }
           next(multiaddrs.shift())
+        }
+
+        conn.on('connect', () => {
+          conn.removeListener('error', connectError)
```

Comment:
maybe we can put this behaviour into a function (adding an error listener and a connect listener to catch connect errors on connecting)